src/lib/LQCALC/LQCALC.h

```diff
@@ -1,23 +1,75 @@
 #pragma once
 
 #include <Arduino.h>
-#include <stdio.h>
 #include "../../src/targets.h"
 
+template <uint8_t N>
 class LQCALC
 {
 public:
-    void init(uint8_t depth);
-    void init();
-    void ICACHE_RAM_ATTR add();
-    void ICACHE_RAM_ATTR inc();
-    uint8_t ICACHE_RAM_ATTR getLQ();
-    void ICACHE_RAM_ATTR reset();
+    LQCALC(void)
+    {
+        reset();
+    }
+
+    /* Set the bit for the current period to true and update the running LQ */
+    void ICACHE_RAM_ATTR add()
+    {
+        LQArray[LQbyte] |= LQmask;
+        LQ += 1;
+    }
+
+    /* Start a new period */
+    void ICACHE_RAM_ATTR inc()
+    {
+        // Increment the counter by shifting one bit higher
+        // If we've shifted out all the bits, move to next idx
+        LQmask = LQmask << 1;
+        if (LQmask == 0)
+        {
+            LQmask = (1 << 0);
+            LQbyte += 1;
+        }
+
+        // At idx N / 32 and bit N % 32, wrap back to idx=0, bit=0
+        if ((LQbyte == (N / 32)) && (LQmask & (1 << (N % 32))))
+        {
+            LQbyte = 0;
+            LQmask = (1 << 0);
+        }
+
+        if ((LQArray[LQbyte] & LQmask) != 0)
+        {
+            LQArray[LQbyte] &= ~LQmask;
+            LQ -= 1;
+        }
+    }
+
+    /* Return the current running total of bits set, in percent */
+    uint8_t ICACHE_RAM_ATTR getLQ() const
+    {
+        return (uint32_t)LQ * 100 / N;
+    }
+
+    /* Initialize and zero the history */
+    void reset()
+    {
+        LQ = 0;
+        LQbyte = 0;
+        LQmask = (1 << 0);
+        for (uint8_t i = 0; i < (sizeof(LQArray)/sizeof(LQArray[0])); i++)
+            LQArray[i] = 0;
+    }
+
+    /*  Return true if the current period was add()ed */
+    bool ICACHE_RAM_ATTR current() const
+    {
+        return (LQArray[LQbyte] & LQmask);
```

Comment:
At the moment current() isnt used any where in the code, which mean LQArray is essentially not necessary.  I wonder if we just use LQ, but then the question is what if it some how increments incorrectly and starts providing miss information.  Or do we keep LQArray and getLQ() performs a sum of the bits?

Im not sure the best way to do this is.  @AlessandroAU @cruwaller ???

Reply:
LQArray is a ring-buffer of the receive validity of the packets. It is tested when rolling over a past position on line 41 which if it was set last round the LQ is decremented and will be incremented in in the add function if a valid packet is received. This allow the running LQ to maintained and not needing to run over the entire array each time the LQ is requested.

Reply:
> which mean LQArray is essentially not necessary
I'm not sure how you arrived at that conclusion. How would you know how many of the last 100 packets were received if you didn't keep track of which packets were received? The LQarray is needed, as pkendall64 said, to know when a missed packet leaves the "last 100 packets" so the LQ can go back up. There is no way to do this with just a simple counter (but you're welcome to prove me wrong).

`current()` isn't used **currently** :hurr: which means the compiler doesn't generate any code for it at all, but it *could* be used to check if a packet was received over the last interval, which means one fewer variable needed for single-missed-packet-based diversity switching. That's the reason I included it. No overhead for including it currently, but the can update the diversity branch after this is accepted for a tiny optimization there.

Reply:
My bad missed that 👍